### DISCUSSION TITLE
pointの付与/取得のAPI仕様について

### DISCUSSION THREAD

**Discussion by ystkfujii** [cloudnativedaysjp/dreamkast#1625] [Closed]
- 現状
  - profileのpointですが、下記のような仕様となっています
    - テーブル設計
      - ユーザのpointを管理するprofilePointテーブル
      - 例）
        | profileId | conference#timestamp | point | reasonId |
        | --------- | -------------------- | ----- | -------- |
        | 1         | cndt2022#11111111    | 10    | 1        |
        | 1         | cndt2022#22222222    | 20    | 1        |
        | 2         | cndt2022#33333333    | 30    | 2        |
    - 登録と取得のAPI実行例
      - シンプルにprofilePointテーブルにレコードを追加/取得する仕様
      ```
      curl -H "Content-Type: application/json" -XPOST -d "{ \"point\": 10,\"reasonId\": 1, \"eventAbbr\": \"cndt2022\" }" https://api.dev.cloudnativedays.jp/api/v1/profile/${PROFILE_ID}/point  
      curl https://api.dev.cloudnativedays.jp/api/v1/profile/${PROFILE_ID}/points?eventAbbr=cndt2022
      ```

ただこの仕様の場合、下記のような問題があるかと思います。
  - 問題１：複数回実行されたときにpointの重複登録が発生する
    - dk側で重複登録を削除する仕組みを入れる方法もありだが、API側でも実装をした方が良さそう
  - 問題２：pointの値がbodyに入っているので、書き換え可能なケースがあるかもしれない
    - bodyにpointを入れているとpointの値をDk側で持つ必要がある
      - それよりはクライアントからpointEventIdを投げさせて、DkはそれをDkFにプロキシする方が楽そう
    - reasonIdを入れるのであればそちらにpointを含めた方が良さそう

そのため、下記のように変更したいと考えています。
- bodyからpointを削除し、pointEvent(旧reason)という概念（？）に含める
  - eventという単語がややこしい気がするので、eventAbbrをconferenceに変更する
  - ここは`他のAPIがeventAbbrなので合わせたい気持ち` vs `eventはややこしいので変えたい気持ち`が私の中で絶賛戦争中です
  - 以前の定例でconferenceにしたいよねという話があったかと思いますので、一旦bodyもconferenceに変更する想定
- profilePointテーブルのprofileIdとpointEventIdの組み合わせが重複登録されたレコードは無視する

具体的には以下のようになります
- テーブル設計に関して
  - 2種類のテーブルを作成する
    - pointEventテーブル
      - pointEventIdは旧reasonId
      - pointを付与するEvent毎およびそのpointを定義するテーブル
      - pointEventIdはユーザの想像のつかないユニークな値
      - 登録は一旦手動
      - 例）
          | conference | pointEventId | point | description   |
          | ---------- | ------------ | ----- | ------------- |
          | cn1        | 1            | 10    | track a mita  |
          | cn1        | 2            | 20    | track b mita  |
          | cn1        | 3            | 30    | offline sanka |
  - profilePointテーブル
    - ユーザのpointを管理するテーブル
    - 例）
        | profileId | conference#timestamp | pointEventId |
        | --------- | -------------------- | ------------ |
        | 1         | cndt2022#11111111    | 1            |
        | 1         | cndt2022#22222222    | 2            |
        | 2         | cndt2022#33333333    | 1            |
- pointの登録処理
  - pointEventIdおよびconferenceを指定してPOSTで実行
  - API実行毎にprofilePointテーブルが作成される
  - 例）：`curl -H "Content-Type: application/json" -XPOST -d "{ \"pointEventId\": 1,\"conference\": \"cndt2022" }" https://api.dev.cloudnativedays.jp/api/v1/profile/${PROFILE_ID}/point`
- pointの取得処理
  - profilePointテーブルのpointEventIdからpointEventテーブルを参照し、pointを取得する
  - 取得時に下記のprofilePointテーブルのレコードは無視する
    - conferenceおよびpointEventIdの組み合わせがpointEventテーブルに存在しないレコード
    - conferenceおよびpointEventIdの組み合わせがprofilePointテーブルに重複して登録されているレコード
  - 例）：`curl https://api.dev.cloudnativedays.jp/api/v1/profile/${PROFILE_ID}/points?conference=cf1`

**Comment by ystkfujii** [comment]
作業途中ですが、このような形で修正予定です
  - https://github.com/cloudnativedaysjp/dreamkast-functions/pull/44

**Reply by ystkfujii** [reply]
ここらへん愚直にループ回してるけど、tsっぽいもう少しいい書き方がありそうな気がしています...

https://github.com/cloudnativedaysjp/dreamkast-functions/blob/63c3c0ee29691c91eb7ea955375015ca106cb622/src/get_profile_point.ts#L72-L86

**Reply by hrk091** [reply]
tsとして良い書き方はわかりませんが、こういう書き方も出来ます

```typescript
    const points: Record<string, profilePointResponse> = {}
    profilePointRecords.Items?.forEach((item) => {
        const ppr = unmarshall(item)
        pointEventRecords.Items?.forEach((v) => {
            const per = unmarshall(v)
            points[per.pointEventId] = points[per.pointEventId] || {
                point: per.point,
                pointEventId: per.pointEventId,
                timestamp: parseInt(ppr['conference#timestamp'].split('#')[1]),
            }
        })
    })

    const total = Object.values(points).reduce((t, v) => t + v.point, 0)
```
